client/tinychain/ml/optimizer.py

```diff
@@ -16,26 +17,44 @@ def optimize(self, i, gradient, inputs, loss):
 
 class GradientDescent(Optimizer):
     @classmethod
-    def create(cls, learning_rate=F32(0.01)):
+    def create(cls, lr=F32(0.01)):
         """Create a new `GradientDescent` optimizer with the given `learning_rate`."""
 
-        return cls({"lr": learning_rate})
+        return cls({"lr": lr})
 
-    def optimize(self, i, gradient, inputs, loss):
+    def optimize(self, i, param_list: List[Parameter]):
         lr = self["lr"]
 
-        def update(gradient, deltas):
-            shape = gradient.shape
+        return [param.value.write(param.value - (param.grad * lr)) for param in param_list]
+
+
+class Adam(Optimizer):
+    @classmethod
+    def create(cls, beta1=F32(0.9), beta2=F32(0.999), lr=F32(1e-3), eps=F32(1e-8)):
+        return cls({'beta1': beta1, 'beta2': beta1, 'lr': lr, 'eps': eps})
+
+    def optimize(self, i, param_list: List[Parameter]):
+        self.m = 
+        self.v = 
+        beta1 = self['beta1']
+        beta2 = self['beta2']
+        lr = self['lr']
+        eps = self['eps']
 
-            if isinstance(shape, dict):
-                return {name: update(gradient[name], deltas[name]) for name in shape}
-            elif isinstance(shape, list) or isinstance(shape, tuple):
-                return [update(gradient[n], deltas[n]) for n in range(len(shape))]
-            else:
-                return gradient.write(gradient - (deltas * lr))
+        self.m = [m.value.write(m.value * beta1 + p.grad * (F32(1) - beta1)) for m, p in zip(self.m, param_list)]
```

Comment:
These layer names are implementation details of a `DNN` which other `Differentiable`s, like a `CNN`, won't share. It would be better to do automatic discovery of the parameters using the `shape` of the model to train. I think that would also remove the need for the `name` parameter in `DNNLayer.load`.

Reply:
Done.
It was a draft version.